R/db-odbc-access.R

```diff
@@ -0,0 +1,240 @@
+                               "ATN(", x, "/ SQR(-", x, "*", x, " + 1)))")                   # Otherwise return ASIN(x) using trig manipulation of ATAN
+                   },
+                   # Inverse hyperbolic sine for the domain x > 1 satisfies log(x + sqrt(x^2 + 1)). x <= 1 returns complex results
+                   asinh         = function(x) {
+                     build_sql("LOG(", x, " + SQR(", x, "^2 + 1))")
+                   },
+                   atan          = sql_prefix("ATN"),
+                   # Note that R takes y then x as arguments
+                   atan2         = function(y, x) {
+                     build_sql("IIF(", y, " > 0, ",
+                                    "IIF(", x, " >= ", y, ", ",
+                                         "ATN(", y, " / ", x, "), ",
+                                         "IIF(", x, " <= -", y, ", ",
+                                              "ATN(", y, " / ", x, "+ ATN(1) * 4), ",
+                                              "ATN(1) * 2 - ATN(", x, " / ", y, "))), ",
+                                    "IIF(", x, ">= -", y, ", ",
+                                         "ATN(", y, "/", x, "), ",
+                                         "IIF(", x, "<=", y, ", ",
+                                              "ATN(", y, "/", x, "- ATN(1)*4), ",
+                                              "-ATN(1)*2 - ATN(", x, "/", y, "))))")
+                   },
+                   # Inverse hyperbolic tan for the domain -1 < x < 1 satisfies log((1 + x) / (1 - x)) / 2.
+                   atanh         = function(x) {
+                     build_sql("LOG((1 + ", x, ") / (1 - ", x, ")) / 2")
+                   },
+                   cot           = function(x) {
+                     build_sql("1 / TAN(", x, ")")
+                   },
+                   coth          = function(x) {
+                     build_sql("(EXP(", x, ") + EXP(-", x, ")) / (EXP(", x, ") - EXP(-", x, "))")
+                   },
+
+
+                   # Strings
+                   nchar         = sql_prefix("LEN"),
+                   tolower       = sql_prefix("LCASE"),
+                   toupper       = sql_prefix("UCASE"),
+                   # Pull `left` chars from the left, then `right` chars from the right to replicate substr
+                   substr        = function(x, start, stop){
+                     right  <- stop - start + 1
+                     left   <- stop
+                     build_sql(
+                       "RIGHT(LEFT(", x, ", ", left, "), ", right, ")")
+                   },
+                   trimws        = sql_prefix("TRIM"),
+                   # No support for CONCAT in Access
+                   paste         = sql_not_supported("paste()"),
+                   paste0        = sql_not_supported("paste0()"),
+
+                   # Logic
+                   # ISNULL() returns -1 for True and 0 for False
```

Comment:
Does access use `-1` for `TRUE` everywhere?

Reply:
Yes, it seems Access tries to mimic a `bit` field, and the two possible integer representations are either `0` for False and `1` for True. 
https://superuser.com/questions/694919/why-do-yes-no-values-in-microsoft-access-map-to-1-and-0
https://stackoverflow.com/questions/8827447/why-is-yes-a-value-of-1-in-ms-access-database

Reply:
In that case, I don't think you should translate to 1 and 0 here. Instead, I think you should define a custom `sql_escape_logical` method.

Reply:
Trying to get my head around this. Are you suggesting this?

```r
is.null       = function(x){
    build_sql("IIF(ISNULL(", x ,"), ", TRUE, ", ", FALSE, ")")
},

is.na         = function(x){
    build_sql("IIF(ISNULL(", x ,"), ", TRUE, ", ", FALSE, ")")
},
```

And a `sql_escape_logical.ACCESS` like:

```r
sql_escape_logical.ACCESS <- function(con, x) {
  y <- as.character(as.integer(x))
  y[is.na(x)] <- "NULL"
  y
}
```

Now this happens:

```r
> ct %>% mutate(x = is.na(speed))
# Source:   lazy query [?? x 3]
# Database: ACCESS 14.00.0000[admin@ACCESS/C:\Users\Administrator\Desktop\R\dbplyr-testing\dplyr-test-db.accdb]
   speed  dist     x
   <dbl> <dbl> <int>
 1    NA     2     1
 2     4    10     0
 3     7     4     0
 4     7    22     0
 5     8    16     0
 6     9    10     0
 7    10    18     0
 8    10    26     0
 9    10    34     0
10    11    17     0
# ... with more rows
```

From a translation of:

```r
> ct %>% mutate(x = is.na(speed)) %>% show_query()
<SQL>
SELECT `speed`, `dist`, IIF(ISNULL(`speed`), 1, 0) AS `x`
FROM `carstest`
```

Reply:
I was thinking

```R
is.null       = sql_prefix("isnull")

sql_escape_logical.ACCESS <- function(con, x) {
  y <- ifelse(x, -1, 0)
  y[is.na(x)] <- "NULL"
  y
}
```

Reply:
I'm just concerned that operations like `x == y` will also yield -1 and 0, so masking for just a few ops is going to be confusing.

Reply:
Ah I see. Now it is always consistent with what Access returns. I changed it and added a test.